src/Common/src/Interop/Windows/sspicli/SecuritySafeHandles.cs

```diff
@@ -1241,6 +1241,120 @@ public new IntPtr DangerousGetHandle()
+
+            int errorCode = (int)Interop.SECURITY_STATUS.InvalidHandle;
+
+            // These are pinned user byte arrays passed along with SecurityBuffers.
+            GCHandle[] pinnedInBytes = null;
+
+            var inUnmanagedBuffer = new Interop.SspiCli.SecBuffer[inSecurityBufferDescriptor.cBuffers];
+            fixed (void* inUnmanagedBufferPtr = inUnmanagedBuffer)
+            {
+                // Fix Descriptor pointer that points to unmanaged SecurityBuffers.
+                inSecurityBufferDescriptor.pBuffers = inUnmanagedBufferPtr;
+                pinnedInBytes = new GCHandle[inSecurityBufferDescriptor.cBuffers];
+                SecurityBuffer securityBuffer;
+                for (int index = 0; index < inSecurityBufferDescriptor.cBuffers; ++index)
+                {
+                    securityBuffer = inSecBuffers[index];
+                    if (securityBuffer != null)
+                    {
+                        inUnmanagedBuffer[index].cbBuffer = securityBuffer.size;
+                        inUnmanagedBuffer[index].BufferType = securityBuffer.type;
+
+                        // Use the unmanaged token if it's not null; otherwise use the managed buffer.
+                        if (securityBuffer.unmanagedToken != null)
+                        {
+                            inUnmanagedBuffer[index].pvBuffer = securityBuffer.unmanagedToken.DangerousGetHandle();
+                        }
+                        else if (securityBuffer.token == null || securityBuffer.token.Length == 0)
+                        {
+                            inUnmanagedBuffer[index].pvBuffer = IntPtr.Zero;
+                        }
+                        else
+                        {
+                            pinnedInBytes[index] = GCHandle.Alloc(securityBuffer.token, GCHandleType.Pinned);
+                            inUnmanagedBuffer[index].pvBuffer = Marshal.UnsafeAddrOfPinnedArrayElement(securityBuffer.token, securityBuffer.offset);
+                        }
+#if TRACE_VERBOSE
+                        if (GlobalLog.IsEnabled)
+                        {
+                            GlobalLog.Print("SecBuffer: cbBuffer:" + securityBuffer.size + " BufferType:" + securityBuffer.type);
+                        }
+#endif
+                    }
+                }
+
+                Interop.SspiCli.CredHandle contextHandle = new Interop.SspiCli.CredHandle();
+                if (refContext != null)
+                {
+                    contextHandle = refContext._handle;
+                }
+                try
+                {
+                    if (refContext == null || refContext.IsInvalid)
+                    {
+                        refContext = new SafeDeleteContext_SECURITY();
+                    }
+
+                    try
+                    {
+                        bool ignore = false;
+                        refContext.DangerousAddRef(ref ignore);
```

Comment:
I agree: there's a lot to optimize here and in the `CompleteAuthToken` which gave me the pattern:
* Remove the extra allocation of a CredHandle
* Refactor the code to avoid duplication
* Remove the AddRef if the refContext was previously null.

I'll leave these for a future change and add them to #3114